### DISCUSSION TITLE
[RuntimeWarning] greenlet.greenlet

### DISCUSSION THREAD

**Discussion by Mat-moran** [Tecnativa/doodba#371] [Q&A] [answered]
Hi,  

I see this in the odoo logs. 

```<frozen importlib._bootstrap>:219: RuntimeWarning: greenlet.greenlet size changed, may indicate binary incompatibility. Expected 144 from C header, got 152 from PyObject```

I read that is related with the greenlet/python version https://stackoverflow.com/questions/64105616/greenlet-runtime-error-and-deployed-app-in-docker-keeps-booting-all-the-workers

```gevent 20.9.0 is required to use (a standard build of) greenlet 0.4.17.``` proposed here --> https://github.com/python-greenlet/greenlet/issues/178#issuecomment-697342964

**It is important?**, is this the right channel for  this question?

I do not know if is for this reason but I am suffering **logouts** and **reentries** constantly.
I see this bucle in my logs constantly. **IS THIS THE NORMAL BEHAVIOUR**?

```
odoo_1    | 2021-01-23 18:42:12,349 6913 INFO ? odoo: Odoo version 14.0
odoo_1    | 2021-01-23 18:42:12,350 6913 INFO ? odoo: Using configuration file at /opt/odoo/auto/odoo.conf
odoo_1    | 2021-01-23 18:42:12,351 6913 INFO ? odoo: addons paths: ['/opt/odoo/custom/src/odoo/odoo/addons', '/var/lib/odoo/addons/14.0', '/opt/odoo/auto/addons']
odoo_1    | 2021-01-23 18:42:12,352 6913 INFO ? odoo: database: odoo@db:5432
odoo_1    | 2021-01-23 18:42:12,612 6913 INFO ? odoo.addons.base.models.ir_actions_report: Will use the Wkhtmltopdf binary at /usr/local/bin/wkhtmltopdf
odoo_1    | <frozen importlib._bootstrap>:219: RuntimeWarning: greenlet.greenlet size changed, may indicate binary incompatibility. Expected 144 from C header, got 152 from PyObject
odoo_1    | <frozen importlib._bootstrap>:219: RuntimeWarning: greenlet.greenlet size changed, may indicate binary incompatibility. Expected 144 from C header, got 152 from PyObject
odoo_1    | <frozen importlib._bootstrap>:219: RuntimeWarning: greenlet.greenlet size changed, may indicate binary incompatibility. Expected 144 from C header, got 152 from PyObject
odoo_1    | <frozen importlib._bootstrap>:219: RuntimeWarning: greenlet.greenlet size changed, may indicate binary incompatibility. Expected 144 from C header, got 152 from PyObject
odoo_1    | <frozen importlib._bootstrap>:219: RuntimeWarning: greenlet.greenlet size changed, may indicate binary incompatibility. Expected 144 from C header, got 152 from PyObject
odoo_1    | <frozen importlib._bootstrap>:219: RuntimeWarning: greenlet.greenlet size changed, may indicate binary incompatibility. Expected 144 from C header, got 152 from PyObject
odoo_1    | <frozen importlib._bootstrap>:219: RuntimeWarning: greenlet.greenlet size changed, may indicate binary incompatibility. Expected 144 from C header, got 152 from PyObject
odoo_1    | <frozen importlib._bootstrap>:219: RuntimeWarning: greenlet.greenlet size changed, may indicate binary incompatibility. Expected 144 from C header, got 152 from PyObject
odoo_1    | <frozen importlib._bootstrap>:219: RuntimeWarning: greenlet.greenlet size changed, may indicate binary incompatibility. Expected 144 from C header, got 152 from PyObject
odoo_1    | 2021-01-23 18:42:13,580 6927 INFO ? odoo: Odoo version 14.0
odoo_1    | 2021-01-23 18:42:13,581 6927 INFO ? odoo: Using configuration file at /opt/odoo/auto/odoo.conf
odoo_1    | 2021-01-23 18:42:13,581 6927 INFO ? odoo: addons paths: ['/opt/odoo/custom/src/odoo/odoo/addons', '/var/lib/odoo/addons/14.0', '/opt/odoo/auto/addons']
odoo_1    | 2021-01-23 18:42:13,581 6927 INFO ? odoo: database: odoo@db:5432
odoo_1    | 2021-01-23 18:42:13,783 6927 INFO ? odoo.addons.base.models.ir_actions_report: Will use the Wkhtmltopdf binary at /usr/local/bin/wkhtmltopdf
odoo_1    | <frozen importlib._bootstrap>:219: RuntimeWarning: greenlet.greenlet size changed, may indicate binary incompatibility. Expected 144 from C header, got 152 from PyObject
odoo_1    | <frozen importlib._bootstrap>:219: RuntimeWarning: greenlet.greenlet size changed, may indicate binary incompatibility. Expected 144 from C header, got 152 from PyObject
odoo_1    | <frozen importlib._bootstrap>:219: RuntimeWarning: greenlet.greenlet size changed, may indicate binary incompatibility. Expected 144 from C header, got 152 from PyObject
odoo_1    | <frozen importlib._bootstrap>:219: RuntimeWarning: greenlet.greenlet size changed, may indicate binary incompatibility. Expected 144 from C header, got 152 from PyObject
odoo_1    | <frozen importlib._bootstrap>:219: RuntimeWarning: greenlet.greenlet size changed, may indicate binary incompatibility. Expected 144 from C header, got 152 from PyObject
odoo_1    | <frozen importlib._bootstrap>:219: RuntimeWarning: greenlet.greenlet size changed, may indicate binary incompatibility. Expected 144 from C header, got 152 from PyObject
odoo_1    | <frozen importlib._bootstrap>:219: RuntimeWarning: greenlet.greenlet size changed, may indicate binary incompatibility. Expected 144 from C header, got 152 from PyObject
odoo_1    | <frozen importlib._bootstrap>:219: RuntimeWarning: greenlet.greenlet size changed, may indicate binary incompatibility. Expected 144 from C header, got 152 from PyObject
odoo_1    | <frozen importlib._bootstrap>:219: RuntimeWarning: greenlet.greenlet size changed, may indicate binary incompatibility. Expected 144 from C header, got 152 from PyObject
```

🙏 help ... .
Thanks and regards,

Odoo version: 14
Postgress version: 13

**Comment by Mat-moran** [comment, marked as answer]
I opened an issue at the OCB repository. 

The problem seems to be that the requirements file in OCB is not updated with gevent version for python > 3.8 and I was aggregating the OCB repository.
https://github.com/OCA/OCB/issues/995

I removed OCB and I only download from Odoo with git-aggregator but **it still installing the wrong gevent version** 
```Requirement already satisfied: gevent in /usr/local/lib/python3.8/site-packages (from odoo==14.0) (1.5.0)```

When I checked the ```requirements.txt``` inside the container in ```custom/src/odoo/```, it is OK with the correct ```gevent version``` for python > 3.8.

what is happening, why it is installing the wrong version of gevent, even if I have it correct now? I did ```docker system prune -a``` to remove all the images in the system and force to rebuild again, I did:
```docker-compose up -d --build``` and I tried also with ```docker-compose buid --no-cache``` it still installing ```gevent=1.5.0``` 

Any ideas of why is still using this version?? **is there other place to specify the version than in the requirements?**

**Reply by pedrobaeza** [reply]
The fix for that is in https://github.com/odoo/odoo/commit/1622aa755bb67e2d9e6efd18366be49b8e4783cb, but it hasn't arrived yet to OCB (tonight will happen).

**Reply by Mat-moran** [reply]
Thanks @pedrobaeza . 

Just to clarify, the requirements are hardcoded to install from OCA/OCB ones in doodba image right? Throw line https://github.com/Tecnativa/doodba/blob/25657f97736fc688045d8b87eff979b14adfcfd7/14.0.Dockerfile#L96 ?

There is a way to circunvent that until OCA/OCB is updated?

**Reply by pedrobaeza** [reply]
It should have landed already, but you can override what you want in pip.txt file in Doodba

**Reply by yajo** [reply]
> the requirements are hardcoded to install from OCA/OCB ones in doodba image right?

You can override that build arg to choose a different odoo flavor, or you can overwrite it in `repos.yaml`. I wouldn't call that "hardcoded". 😋
